libraries/AC_Avoidance/AC_Avoid.cpp

```diff
@@ -223,25 +223,51 @@ void AC_Avoid::adjust_velocity(Vector3f &desired_vel_cms, bool &backing_up, floa
         // this has to be done for x,y,z seperately. For eg, user is doing fine in "x" direction but might need backing up in "y".
         if (!is_zero(desired_backup_vel.x)) {
             if (is_positive(desired_backup_vel.x)) {
-                desired_vel_cms.x = MAX(desired_vel_cms.x, desired_backup_vel.x);
+                desired_vel_cms.x = MAX(desired_vel_cms_original.x, desired_backup_vel.x);
```

Comment:
This does not look correct to me. What if the original velocity was "unsafe" and will lead to an obstacle collision? Same applies for the changes down below

Reply:
```
if (!is_zero(desired_backup_vel.x)) {
            if (is_positive(desired_backup_vel.x)) {
                desired_vel_cms.x = MAX(desired_vel_cms_original.x, desired_backup_vel.x);
            } else {
                desired_vel_cms.x = MIN(desired_vel_cms_original.x, desired_backup_vel.x);
            }
        }
```
let's take this example..... if desired_backup_vel is positive then direction of escaping "backup" is +ve then you can take the max positive safely.    if desired_backup_vel is negative then direction of escaping "backup" is -ve then again you can take the max "numerically minimum " value safely.

Reply:
Now the second part:

```
if (!_vector_to_obstacle_inaction.is_zero() && desired_backup_vel.is_zero())
    {
        bool sign_desired_vel_cms_original_x = desired_vel_cms_original.x>=0.0f;
        bool sign__vector_to_obstacle_inaction_x = _vector_to_obstacle_inaction.x >=0.0f;
        
        bool sign_desired_vel_cms_original_y = desired_vel_cms_original.y>=0.0f;
        bool sign__vector_to_obstacle_inaction_y = _vector_to_obstacle_inaction.y >=0.0f;
        
        bool sign_desired_vel_cms_original_z = desired_vel_cms_original.z>=0.0f;
        bool sign__vector_to_obstacle_inaction_z = _vector_to_obstacle_inaction.z >=0.0f;
        
        if (sign_desired_vel_cms_original_x != sign__vector_to_obstacle_inaction_x)
        {
            desired_vel_cms.x = desired_vel_cms_original.x;   
        }

        if (sign_desired_vel_cms_original_y != sign__vector_to_obstacle_inaction_y)
        {
            desired_vel_cms.x = desired_vel_cms_original.x;   
        }

        if (sign_desired_vel_cms_original_z != sign__vector_to_obstacle_inaction_z)
        {
            desired_vel_cms.x = desired_vel_cms_original.x;   
        }
    }
```

whenever a vehicle decides to stop then again you can move away from the obstacle safely if the user wants to.

Reply:
@HefnySco just because the direction of desired_backup_vel is positive in one direction, does not mean that direction is safe. Imagine it this way, obstacles are surrounded by both +ve x and -ve x directions (say we have a 360-degree lidar), but the obstacle in the +ve direction is very very close (much closer than the obstacle at -ve x end) so we have to back up in the -ve direction. With your changes, we'll back up with full speed in the negative direction even though there is still an obstacle present there. If on the other hand, there is no obstacle in one direction, then the velocity is anyway not changed so then desired_vel_cms_original = desired_vel_cms.

Reply:
Just to be clear, I am not saying you haven't found a real bug/problem. I just can't understand how this PR at the moment is the correct solution to it

Reply:
> @HefnySco just because the direction of desired_backup_vel is positive in one direction, does not mean that direction is safe. Imagine it this way, obstacles are surrounded by both +ve x and -ve x directions (say we have a 360-degree lidar), but the obstacle in the +ve direction is very very close (much closer than the obstacle at -ve x end) so we have to back up in the -ve direction. With your changes, we'll back up with full speed in the negative direction even though there is still an obstacle present there. If on the other hand, there is no obstacle in one direction, then the velocity is anyway not changed so then desired_vel_cms_original = desired_vel_cms.

I agree with you .... I will dig deeper and repeat the scenario... I believe the key is in the bottom layer functions "find_max_quadrant_velocity". I will check.

Many thanks for your feedback.